statsmodels/base/model.py

```diff
@@ -796,20 +796,16 @@ def predict(self, exog=None, transform=True, *args, **kwargs):
 
         exog_index = exog.index if _is_using_pandas(exog, None) else None
 
-        if transform and hasattr(self.model, 'formula') and exog is not None:
+        if transform and hasattr(self.model, 'formula') and (exog is not None):
             from patsy import dmatrix
-            exog = pd.DataFrame(exog)  # user may pass series, if one predictor
-            if exog_index is None:  # user passed in a dictionary
-                exog_index = exog.index
+            if isinstance(exog, pd.Series):
+                exog = pd.DataFrame(exog)
             exog = dmatrix(self.model.data.design_info.builder,
                            exog, return_type="dataframe")
-            if len(exog) < len(exog_index):
+            if (exog_index is not None) and (len(exog) < len(exog_index)):
                 # missing values, rows have been dropped
-                if exog_index is not None:
-                    exog = exog.reindex(exog_index)
-                else:
-                    import warnings
-                    warnings.warn("nan rows have been dropped", ValueWarning)
```

Comment:
from what I can tell staring at these changes
The only real change for pre-existing cases is that we loose this warning.

If I read this now correctly, then this case was empty. i.e. if exog_index is None, then the if len(exog_index) would already raise an exception.

Reply:
I think we can get the warning back if we keep len(exog) before calling dmatrix and compare it with len(exog) after dmatrix.
This would keep the current code but change the if in old line 809.

I guess this could be added to base test_generic TestWaldAnovaOLSF.test_predict_missing
(I ran the unit test after changing predict so always an ndarray is returned, and it causes 8 test failures in base test_generic. So coverage is there, but not for a warning.)
Actually, I think your new test case, could check for a warning that we loose a row that patsy deleted.